src/eigenvalues.js

```diff
@@ -1,115 +1,37 @@
-        const hdy = sum(sdy, (v, i) => v * h[i]);
-        const dx = sdx.map((v, i) => 2 * (v - hdx * h[i]));
-        const dy = sdy.map((v, i) => 2 * (v - hdy * h[i]));
-        
-        for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) {
-            m[idx(x, y)] -= h[y] * dy[x] + h[x] * dx[y];
-        }
-    }
-    return m;
-}
-
-// returns upper triangler matrix: m[y][x] == 0 when y >= x + 1
-function qr(mat) {
-    const n = Math.sqrt(mat.length);
-    const m = Array.from(mat);
-    console.assert(Number.isInteger(n));
-    const idx = (x, y) => y * n + x;
-    let k = n;
-
-    while (k >= 2) {
-        if (Math.abs(m[idx(k - 2, k - 1)]) < Number.EPSILON) {
-            k--;
-            continue;
-        }
-        
-        // eigenvalue of last 2x2 sub matrix: l^2 - tr * l + det = 0
-        const a = m[idx(k - 1, k - 1)], b = m[idx(k - 2, k - 1)],
-              c = m[idx(k - 1, k - 2)], d = m[idx(k - 2, k - 2)];
-        const tr = a * d, det = a * d - b * c;
-        const disc = Math.sqrt(tr * tr - 4 * det) || 0;
-        const l1 = (tr + disc) / 2, l2 = (tr - disc) / 2;
-        const mu = a - (Math.abs(l1) < Math.abs(l2) ? l1 : l2);
-        // (option) pre process M = M - mu * I 
-        for (let i = 0; i < k; i++) m[idx(i, i)] -= mu;
-        
-        // init q as unit matrix
-        const idxq = (x, y) => y * k + x;
-        const q = Array(k * k).fill(0);
-        for (let i = 0; i < k; i++) q[idxq(i, i)] = 1;
-        
-        // rotate to makes M => Q * R (R store to M)
-        for (let i = 0; i < k - 1; i++) {
-            const a1 = m[idx(i, i)], a2 = m[idx(i, i + 1)]; 
-            const base = Math.hypot(a1, a2);
-            const cos = base < Number.EPSILON ? 0 : a1 / base;
-            const sin = base < Number.EPSILON ? 0 : a2 / base;
-            // make R
-            m[idx(i, i)] = base;
-            m[idx(i, i + 1)] = 0;
-            for (let x = i + 1; x < k; x++) {
-                const e1 = m[idx(x, i)], e2 = m[idx(x, i + 1)];
-                m[idx(x, i)] = e1 * cos + e2 * sin;
-                m[idx(x, i + 1)] = e2 * cos - e1 * sin;
-            }
-            // make Q
-            for (let y = 0; y < k; y++) {
-                const e1 = q[idxq(i, y)], e2 = q[idxq(i + 1, y)];
-                q[idxq(i, y)] = e1 * cos + e2 * sin;
-                q[idxq(i + 1, y)] = e2 * cos - e1 * sin;
-            }
-        }
 
-        // next M as R * Q
-        for (let y = 0; y < k; y++) {
-            const ry = Array.from(Array(k - y), (_, j) => m[idx(y + j, y)]);
-            for (let x = 0; x < k; x++) {
-                m[idx(x, y)] = sum(ry, (v, j) => v * q[idxq(x, j + y)]);
-            }
-        }
-        
-        // (option) post process M = M + mu * I
-        for (let i = 0; i < k; i++) m[idx(i, i)] += mu;        
+function powerIteration(m, n = 100, error = 1e-6) {
+  const max_m = max(m)
+  let x = Array(m[0].length).fill(0).map(() => { return random(1, max_m) });
```

Comment:
is it worth adding the reason for this here (I had to look it up - looks like we want to decrease the chance we pick an orthogonal vector, which must be vanishingly unlikely with a matrix of this size)